### PR TITLE
Change beam exit condition

### PR DESCRIPTION
**Pull request by vince62s**
@funboarder13920 please check the very last test_beam_search in test_beam_lm_update_memory_length_when_finished(self), because I made a change without knowing exactly why this PR requires this change.
Thanks.

### CODE REVIEW

**Comment by funboarder13920** [comment]
Thanks for the split of the PR.
Can you explain the change you made to the test files ?
Can you write a test to force a beam to end with a simulation of EOS ? Currently it doesn't look like we understand the steps required to trigger a beam ending.

File: onmt/tests/test_beam_search.py [hunk 1]
```diff
@@ -292,10 +293,11 @@ def test_beam_is_done_when_n_best_beams_eos_using_min_length(self):
             else:
                 word_probs[0::beam_sz, eos_idx] = valid_score_dist[0]
                 word_probs[1::beam_sz, eos_idx] = valid_score_dist[0]
+                word_probs[2::beam_sz, eos_idx] = valid_score_dist[0]
                 # provide beam_sz other good predictions in other beams
                 for k, (j, score) in enumerate(
                         zip(_non_eos_idxs, valid_score_dist[1:])):
-                    beam_idx = min(beam_sz - 1, k)
+                    beam_idx = min(beam_sz - 2, k)
```
Comment (funboarder13920) [inline onmt/tests/test_beam_search.py:300]: Why is this change required ?

File: onmt/tests/test_beam_search.py [hunk 2]
```diff
@@ -617,5 +622,6 @@ def test_beam_lm_update_memory_length_when_finished(self):
         self.finish_first_beam_step(beam)
 
         n_steps = beam.alive_seq.shape[-1] - 1
+
         self.assertTrue(beam.memory_lengths.equal(
-            n_steps+fn_map_state(src_lengths[1:], dim=0)))
+            n_steps+fn_map_state(src_lengths, dim=0)))
```
Comment (funboarder13920) [inline onmt/tests/test_beam_search.py:627]: The purpose of the test is to check everything is ok when all the beam of an example end. Here it does not happen.
If the method finish_first_beam_step I force all the scores of the beams of eos for the first example. Why is the beam not ending then ?

File: onmt/translate/beam_search.py [hunk 3]
```diff
@@ -189,18 +189,18 @@ def update_finished(self):
                     self.is_finished[i].all()
             else:
                 finish_flag = self.top_beam_finished[i] != 0
-            if finish_flag and len(self.hypotheses[b]) >= self.n_best:
+            if finish_flag and len(self.hypotheses[b]) >= self.beam_size:
                 best_hyp = sorted(
-                    self.hypotheses[b], key=lambda x: x[0], reverse=True)
+                    self.hypotheses[b], key=lambda x: x[0],
```
Comment (funboarder13920) [inline onmt/translate/beam_search.py:194]: Could everything after the if statement be moved in a method with a clear name (register_hypotheses for ex) ?

File: onmt/tests/test_beam_search.py (hunk 2 above)
Comment (vince62s) [reply onmt/tests/test_beam_search.py]: This is not true. Out of the BEAM_SZ=5 beams only 3 out of 5 finish, which means that when the exit condition is ">=n_best" with n_best>=4, according to before this PR, or " >=beam_size" taking into account this PR, then no example is completely finished, and none are removed, hence memory_lengths still contains 3 examples.
This test, as well as some others are not 100% bullet proof wrt the existing code.

File: onmt/translate/beam_search.py (hunk 3 above)
Comment (vince62s) [reply onmt/translate/beam_search.py:194]: in another PR please ....

**Comment by vince62s** [comment]
All tests are ok and should be more bullet proof to changes in n_best, beam_size ....

File: onmt/tests/test_beam_search.py [hunk 4]
```diff
@@ -502,11 +492,11 @@ def second_step(self, beam, expected_beam_scores, expected_len_pen):
     def third_step(self, beam, expected_beam_scores, expected_len_pen):
         # assumes beam 0 finished on last step
         scores_3 = torch.log_softmax(torch.tensor(
-            [[0, 0, 5000, 0, 5000, .51, .2, 0],  # beam 0 shouldn't cont
+            [[0, 0, 10000, 0, 5000, .51, .2, 0],  # beam 0 shouldn't cont
              [0, 0, 0, 0, 0, 0, 0, 0],
-             [0, 0, 0, 0, 0, 5000, 0, 0],
-             [0, 0, 0, .2, .2, .2, .2, .2],
-             [0, 0, 50, 0, .2, .2, .2, .2]]  # beam 4 -> beam 1 should die
+             [0, 0, 10000, 0, 0, 5000, 0, 0],
```
Comment (funboarder13920) [inline onmt/tests/test_beam_search.py:497]: Why is this change required ?
Same for the one next line

File: onmt/tests/test_beam_search.py [hunk 5]
```diff
@@ -526,11 +516,10 @@ def third_step(self, beam, expected_beam_scores, expected_len_pen):
             expected_beam_scores / expected_len_pen))
         self.assertTrue(beam.topk_ids.equal(expected_preds_3))
         self.assertTrue(beam.current_backptr.equal(expected_bptr_3))
-        self.assertEqual(beam.is_finished.sum(), self.BATCH_SZ)
-        # new beam 1 finished
-        self.assertTrue(beam.is_finished[:, 1].all())
-        # new beam 1 is old beam 4
-        self.assertTrue(expected_bptr_3[:, 1].eq(4).all())
+        # we finish 3 hyps per example in this step
```
Comment (funboarder13920) [inline onmt/tests/test_beam_search.py:519]: Why were the test changed ?

File: onmt/tests/test_beam_search.py [hunk 6]
```diff
@@ -290,13 +290,8 @@ def test_beam_is_done_when_n_best_beams_eos_using_min_length(self):
...
             else:
-                word_probs[0::beam_sz, eos_idx] = valid_score_dist[0]
-                word_probs[1::beam_sz, eos_idx] = valid_score_dist[0]
-                # provide beam_sz other good predictions in other beams
-                for k, (j, score) in enumerate(
-                        zip(_non_eos_idxs, valid_score_dist[1:])):
-                    beam_idx = min(beam_sz - 1, k)
-                    word_probs[beam_idx::beam_sz, j] = score
+                for j in range(beam_sz):
+                    word_probs[j::beam_sz, eos_idx] = valid_score_dist[0]
```
Comment (funboarder13920) [inline onmt/tests/test_beam_search.py:294]: I think the purpose of the test is not the same anymore

File: onmt/tests/test_beam_search.py (hunk 4 above)
Comment (vince62s) [reply onmt/tests/test_beam_search.py:497]: The test was designed to have exactly n_best = 3 beams to finish, we need more beams to finish to satisfy the exit condition.
Before we had:
1 beam finishing in first step
1 beam finishing in second step
1 beam finishing in third step
total = 3
Now we need 5 beams to finish so I am changing the third step so that 3 beams finish in this step.

File: onmt/tests/test_beam_search.py (hunk 5 above)
Comment (vince62s) [reply onmt/tests/test_beam_search.py:519]: As stated above we no longer test the same thing given that 3 beams finish instead of 1 in this step

File: onmt/tests/test_beam_search.py (hunk 6 above)
Comment (vince62s) [reply onmt/tests/test_beam_search.py:294]: well yes and no, but just the title of the test is no longer relevant "done when n_best beams" which is no longer the reality since we need to "beam.done" when we have beam size hyps. This is the reason why I force all beams to finish in tihs place.

Comment (francoishernandez) [reply onmt/tests/test_beam_search.py:294]: Shall we change the name of the test then? :)

File: onmt/tests/test_beam_search.py (hunk 5 above)
Comment (funboarder13920) [reply onmt/tests/test_beam_search.py:519]: what about self.assertTrue(beam.is_finished[:, 1].all()) ?

Comment (vince62s) [reply onmt/tests/test_beam_search.py:519]: but then we can also add:
self.assertTrue(beam.is_finished[:, 0].all())
self.assertTrue(beam.is_finished[:, 2].all())
to me the most important thing is to make sure beam.done is true, specifying the 3 that died is not relevant, but if you insist then we need to add the 3 of them.

**Comment by francoishernandez** [comment]
Let's wrap up on this. Let's change the name of the modified test and merge.
We can always make a new pass on this later.
